dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java

```diff
@@ -174,6 +183,88 @@ public BooleanFormula encodeMemory(SolverContext ctx) {
         return fmgr.getBooleanFormulaManager().and(addrExprs);
     }
 
+    /**
+     * @param x
+     * Some event of a program to be encoded.
+     * @param y
+     * Another event of the same program.
+     * @param exec
+     * Analysis performed on the associated program.
+     * @param ctx
+     * Builder of expressions and formulas.
+     */
+    public static BooleanFormula execution(Event x, Event y, ExecutionAnalysis exec, SolverContext ctx) {
+        if(x.exec()==y.exec() || exec.isImplied(x,y)) {
+            return x.exec();
+        }
+        if(exec.isImplied(y,x)) {
+            return y.exec();
+        }
+        BooleanFormulaManager bmgr = ctx.getFormulaManager().getBooleanFormulaManager();
+        return x.getCId() < y.getCId() ? bmgr.and(x.exec(), y.exec()) : bmgr.and(y.exec(), x.exec());
```

Comment:
`and` is reflexive, thus this conditional is useless

Reply:
I think you mean commutative. That is in principle true, but we actually need to check if that makes a difference in performance in any way, because it can simplify detection of identical subterms.

I just noticed, that you took the code from `Relation.getExecVar` and adapted it.
I think more maximal sharing, you still might want to order `x` and `y` by `cId` first.
There is this hypothetical case where both `x` and `y` imply each other but are not merged (i.e. their `exec`-variables are different). This case shouldn't happen as long as we merge `exec`-vars though.

Reply:
Yes, I meant commutative.

I don't get you following comment. Adding `A /\ B` or `B /\ A` to the solver should be exactly the same. The constraints are equivalent and the solver should decide internally how to rewrite this in whatever way it is more efficient for it. Deciding the order of terms should not be the responsibility of the user. Otherwise we will have thousend of decisions all over the code.

Reply:
I agree that **simplification** does improve solver performance, but here we are talking about equivalent constraints with the same number of sub-terms (more a reordering rewriting than a simplification). If it is really true that the solver performance would be better if we add `A /\ B` instead of `B /\ A`, then where to draw the line? Shall we add `programEncoder /\ wmmEncoder` or `wmmEncoder /\ programEncoder` ... this looks like a rabbit hole to me.

Reply:
I think you misunderstood the intention here. It is not about `A /\ B` vs. `B /\ A` but it is about deciding on one of them and sticking to it throughout the whole formula. If you don't, then you find both variations in the same formula and depending on the solver, he might not be able to identify both expressions as equivalent (since they are syntactically distinct). 
The general question of which order is the better one is not answerable, it heavily depends on the actual solver and the actual instance you look at. But I think it is generally true that it is better to avoid asserting different permutations of the same formula because the solver might fail to identify them as equivalent.